File: BFS/Binary Tree Level Order Traversal II/Binary Tree Level Order Traversal II.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> res;
    vector<int> v;
    stack<int> bottomBFS;
    queue<TreeNode*> q;
    void BFS(TreeNode* root){
        int level = 0;
        q.push(root);
        while(!q.empty()){
            level++;
            int n = q.size();
            while(n--){
                TreeNode* temp = q.front();
                q.pop();
                v.push_back(temp->val);
                if(temp->left) q.push(temp->left);
                if(temp->right) q.push(temp->right);
            }
            res.push_back(v);
            v.clear();
        }
      // reverse 2D vector after we got all nodes of BFS.
        reverse(res.begin(), res.end());
    }
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        if(root == NULL) return res;
        BFS(root);
        return res;
    }
};

File: BFS/Binary Tree Level Order Traversal II/Binary Tree Level Order Traversal II.java

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.*;
import java.util.Collections;
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        List<List<Integer>> res = new ArrayList<>();
        if(root == null ) return res;
        q.offer(root);
        while(!q.isEmpty()){
            int size = q.size();
            List<Integer> list = new ArrayList<>();
            for(int i = 0; i < size; i++){
                if(q.peek().left != null)
                    q.offer(q.peek().left);
                if(q.peek().right != null)
                    q.offer(q.peek().right);
                list.add(q.poll().val);
            }
            res.add(list);
        }
      Collections.reverse(res);
       return res;
    }
}
